lib/kube/proxy/forwarder.go

```diff
@@ -2148,3 +2312,358 @@ func (f *Forwarder) removeKubeDetails(name string) {
+		return nil, trace.Wrap(err)
+	}
+	// status holds the returned response code.
+	var status int
+	// Check if the target Kubernetes cluster is not served by the current service.
+	// If it's the case, forward the request to the target Kube Service where the
+	// filtering logic will be applied.
+	if !f.isLocalKubeCluster(sess) {
+		rw := httplib.NewResponseStatusRecorder(w)
+		sess.forwarder.ServeHTTP(rw, req)
+		status = rw.Status()
+	} else {
+		memoryRW := responsewriters.NewMemoryResponseWriter()
+		listReq := req.Clone(req.Context())
+		// reset body and method since list does not need the body response.
+		listReq.Body = nil
+		listReq.Method = http.MethodGet
+		_, err = f.listPods(ctx, memoryRW, listReq, p)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		status, err = f.handleDeleteCollectionReq(req, &sess.authContext, memoryRW, w)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	if sess.noAuditEvents {
+		return nil, nil
+	}
+	// Emit audit event.
+	event := &apievents.KubeRequest{
+		Metadata: apievents.Metadata{
+			Type: events.KubeRequestEvent,
+			Code: events.KubeRequestCode,
+		},
+		UserMetadata: ctx.eventUserMeta(),
+		ConnectionMetadata: apievents.ConnectionMetadata{
+			RemoteAddr: req.RemoteAddr,
+			LocalAddr:  sess.kubeAddress,
+			Protocol:   events.EventProtocolKube,
+		},
+		ServerMetadata: apievents.ServerMetadata{
+			ServerID:        f.cfg.HostID,
+			ServerNamespace: f.cfg.Namespace,
+		},
+		RequestPath:               req.URL.Path,
+		Verb:                      req.Method,
+		ResponseCode:              int32(status),
+		KubernetesClusterMetadata: ctx.eventClusterMeta(),
+	}
+	r := parseResourcePath(req.URL.Path)
+	if r.skipEvent {
+		return nil, nil
+	}
+	r.populateEvent(event)
+	if err := f.cfg.AuthClient.EmitAuditEvent(f.ctx, event); err != nil {
+		f.log.WithError(err).Warn("Failed to emit event.")
+	}
+
+	return nil, nil
+}
+
+func (f *Forwarder) handleDeleteCollectionReq(req *http.Request, authCtx *authContext, memWriter *responsewriters.MemoryResponseWriter, w http.ResponseWriter) (int, error) {
+	const internalErrStatus = http.StatusInternalServerError
+	// get content-type value
+	contentType := responsewriters.GetContentHeader(memWriter.Header())
+	encoder, decoder, err := newEncoderAndDecoderForContentType(contentType, newClientNegotiator())
+	if err != nil {
+		return internalErrStatus, trace.Wrap(err)
+	}
+
+	deleteOptions, err := parseDeleteCollectionBody(req.Body, decoder)
+	if err != nil {
+		return internalErrStatus, trace.Wrap(err)
+	}
+	req.Body.Close()
+
+	// decode memory rw body.
+	obj, err := decodeAndSetGVK(decoder, memWriter.Buffer().Bytes())
+	if err != nil {
+		return internalErrStatus, trace.Wrap(err)
+	}
+
+	details, err := f.findKubeDetailsByClusterName(authCtx.kubeCluster)
+	if err != nil {
+		return internalErrStatus, trace.Wrap(err)
+	}
+	switch o := obj.(type) {
+	case *metav1.Status:
+		// Do nothing.
```

Comment:
Why are Status objects expected here? Is there always a status in the list?

Reply:
Status objects exist when Kube API returns an error such as missing permissions.